dev/io.openliberty.restfulWS.internal.logging.filter/src/io/openliberty/restfulWS/internal/logging/filter/LoggingFilter.java

```diff
@@ -0,0 +1,157 @@
+ * and the outbound HTTP response from the server. 
+ */
+public class LoggingFilter implements ContainerRequestFilter, ContainerResponseFilter {
+    private final static TraceComponent tc = Tr.register(LoggingFilter.class);
+    private final static String LS = System.lineSeparator();
+    private final static String REQUEST_TEMPLATE = Tr.formatMessage(tc, "HTTP_REQUEST") + LS
+                                                 + "> %s %s" + LS + LS  // HTTP method + path
+                                                 + "%s" + LS + LS       // Headers
+                                                 + "%s";                // HTTP body
+
+    private final static String RESPONSE_TEMPLATE = Tr.formatMessage(tc, "HTTP_REQUEST") + LS 
+                                                  + "< %d %s" + LS + LS // HTTP status code + description
+                                                  + "%s" + LS + LS      // Headers
+                                                  + "%s";               // HTTP body
+
+    private final static String UNLOGGED = Tr.formatMessage(tc, "UNLOGGED");
+    private final static String NO_ENTITY = Tr.formatMessage(tc, "NO_ENTITY");
+    private final boolean LOG_ENTITY = AccessController.doPrivileged((PrivilegedAction<Boolean>) () -> 
+        Boolean.getBoolean("io.openliberty.restfulWS.logging.filter.logHTTPEntities"));
+    
+
+    @Context
+    Providers providers;
+
+    @Override
+    public void filter(ContainerRequestContext reqCtx) throws IOException {
+        final String httpMethod = reqCtx.getMethod();
+        StringBuilder sb = new StringBuilder();
+        UriInfo uriInfo = reqCtx.getUriInfo();
+        sb.append(uriInfo.getBaseUri()).append(" ").append(uriInfo.getPath());
+        MultivaluedMap<String, String> queryParams = uriInfo.getQueryParameters();
+        if (!queryParams.isEmpty()) {
+            sb.append("?");
+            for (Entry<String, List<String>> e : queryParams.entrySet()) {
+                for (String value : e.getValue()) {
+                    sb.append(e.getKey()).append("=").append(value).append("&");
+                }
+            }
+            sb.deleteCharAt(sb.length()-1); //remove last ampersand
+        }
+        final String path = sb.toString();
+
+        sb = new StringBuilder();
+        for (String headerName : reqCtx.getHeaders().keySet()) {
+            sb.append(headerName).append(": ").append(reqCtx.getHeaderString(headerName)).append(LS);
+        }
+        final String headers = sb.toString();
+
+        // it's expensive to read the HTTP entity (requires copying the stream - and the full stream onto the heap)
+        // so best to avoid unless user really wants to see it:
+        final String entity;
+        if (!LOG_ENTITY) {
+            entity = UNLOGGED;
+        } else if (!reqCtx.hasEntity()) {
+            entity = NO_ENTITY;
+        } else {
+            byte[] entityBytes = toByteArray(reqCtx.getEntityStream());
```

Comment:
Uh, not 100% sure, but I'm worried about this causing the entity stream to be read when the backend tries to read it. Have you tested this?

Reply:
yes, this will read the entity stream - which is why I have it set to not read/print it by default.  I tested it to verify that the resource will still be able to read the stream (from the newly created ByteArrayInputStream in the code below). It's not great for performance, but I think users would only enable this if they were willing to give up some performance for additional logging.